ModbusClientMixin.write_registers typing could be loosened

<!--
Before opening a new issue, make sure you do the following:

- Check that your issue isn't already filed: https://github.com/pymodbus-dev/pymodbus/issues
- Check the discussions forum https://github.com/pymodbus-dev/pymodbus/discussions
- Prepare a short, runnable example that reproduce the issue with the latest development version of Pymodbus
-->

Note: all code is untested except by unit tests.

### Pymodbus Specific

- Client

### Description

<!-- What were you trying, what has happened, what went wrong, and what did you expect? -->

The typed interface for `ModbusClientMixin.write_registers` has flip-flopped a few times in recent releases, but it's still slightly wrong IMO.

The current typing, as of the time of writing, is:

```python
    def write_registers(
        self,
        address: int,
        values: list[bytes | int],
        *,
        slave: int = 1,
        no_response_expected: bool = False
    ) -> T:
```

The problem with this typing is that any code calling into this function that is explicitly typed as `values: Sequence[int]` or `values: list[int]` will fail type checks.

It seems the motivation to migrate this from `list[int]` to `list[int] | list[bytes]` to `list[int|bytes]` is to denote that the input can be a mixture of ints or bytes and that the interface will handle conversions as necessary.

Unfortunately, because `list` is mutable, it is invariant. Type checkers, like mypy, assume because it's typed as `list[int|byte]` that the function could _potentially_ mutate this list to contain a type that a caller, who has typed their interfaces to be `list[int]` to be unaware of and would be unable to handle.

These interfaces _should not_ mutate the value, so can and probably should loosen their types as much as possible... with a few tweak, it could be as generic as:

```python
values: Iterable[int|bytes]
```

The underlying code really only needs it to be iterable (via `enumerate`). The only exception is:

```python
self.count: int = count or len(registers)
```

Which would mean it needs to be `Container` or `Sequence` however I think this could be avoided by simply doing:

```python
self.count: int = count or len(self.registers)
```

### Code and Logs

```diff
diff --git a/examples/client_async_calls.py b/examples/client_async_calls.py
index d2c118e5..2c119e8d 100755
--- a/examples/client_async_calls.py
+++ b/examples/client_async_calls.py
@@ -163,13 +163,15 @@ async def async_write_registers_mypy(client):
     await client.write_registers(1, regs1, slave=SLAVE)
     rr = await client.read_holding_registers(1, count=len(regs1), slave=SLAVE)
     assert not rr.isError()  # test that call was OK
+    assert rr.registers is not regs1
     assert rr.registers == regs1
 
     regs2: list[bytes] = [b'\x01\x02', b'\x03\x04']
     await client.write_registers(1, regs2, slave=SLAVE)
     rr = await client.read_holding_registers(1, count=len(regs2), slave=SLAVE)
     assert not rr.isError()  # test that call was OK
-    assert rr.registers == regs2
+    assert rr.registers is not regs2
+    assert rr.registers == [int.from_bytes(i, byteorder="big") for i in regs2]
 
 
 async def async_handle_input_registers(client):
diff --git a/pymodbus/client/mixin.py b/pymodbus/client/mixin.py
index 23eebc89..215e7353 100644
--- a/pymodbus/client/mixin.py
+++ b/pymodbus/client/mixin.py
@@ -2,6 +2,7 @@
 from __future__ import annotations
 
 import struct
+from collections.abc import Iterable
 from enum import Enum
 from typing import Generic, TypeVar
 
@@ -501,7 +502,7 @@ class ModbusClientMixin(Generic[T]):  # pylint: disable=too-many-public-methods
     def write_registers(
         self,
         address: int,
-        values: list[bytes | int],
+        values: Iterable[bytes | int],
         *,
         slave: int = 1,
         no_response_expected: bool = False
(venv) vfazio@vfazio4 /mnt/development/pymodbus $ PAGER= git diff
diff --git a/examples/client_async_calls.py b/examples/client_async_calls.py
index d2c118e5..2c119e8d 100755
--- a/examples/client_async_calls.py
+++ b/examples/client_async_calls.py
@@ -163,13 +163,15 @@ async def async_write_registers_mypy(client):
     await client.write_registers(1, regs1, slave=SLAVE)
     rr = await client.read_holding_registers(1, count=len(regs1), slave=SLAVE)
     assert not rr.isError()  # test that call was OK
+    assert rr.registers is not regs1
     assert rr.registers == regs1
 
     regs2: list[bytes] = [b'\x01\x02', b'\x03\x04']
     await client.write_registers(1, regs2, slave=SLAVE)
     rr = await client.read_holding_registers(1, count=len(regs2), slave=SLAVE)
     assert not rr.isError()  # test that call was OK
-    assert rr.registers == regs2
+    assert rr.registers is not regs2
+    assert rr.registers == [int.from_bytes(i, byteorder="big") for i in regs2]
 
 
 async def async_handle_input_registers(client):
diff --git a/pymodbus/client/mixin.py b/pymodbus/client/mixin.py
index 23eebc89..215e7353 100644
--- a/pymodbus/client/mixin.py
+++ b/pymodbus/client/mixin.py
@@ -2,6 +2,7 @@
 from __future__ import annotations
 
 import struct
+from collections.abc import Iterable
 from enum import Enum
 from typing import Generic, TypeVar
 
@@ -501,7 +502,7 @@ class ModbusClientMixin(Generic[T]):  # pylint: disable=too-many-public-methods
     def write_registers(
         self,
         address: int,
-        values: list[bytes | int],
+        values: Iterable[bytes | int],
         *,
         slave: int = 1,
         no_response_expected: bool = False
diff --git a/pymodbus/pdu/pdu.py b/pymodbus/pdu/pdu.py
index b65426c6..67a9fdd6 100644
--- a/pymodbus/pdu/pdu.py
+++ b/pymodbus/pdu/pdu.py
@@ -4,7 +4,7 @@ from __future__ import annotations
 import asyncio
 import struct
 from abc import abstractmethod
-from collections.abc import Sequence
+from collections.abc import Iterable
 from enum import Enum
 from typing import cast
 
@@ -26,7 +26,7 @@ class ModbusPDU:
             address: int = 0,
             count: int = 0,
             bits: list[bool] | None = None,
-            registers: Sequence[int | bytes] | None = None,
+            registers: Iterable[int | bytes] | None = None,
             status: int = 1,
         ) -> None:
         """Initialize the base data for a modbus request."""
@@ -36,11 +36,16 @@ class ModbusPDU:
         self.bits: list[bool] = bits or []
         if not registers:
             registers = []
-        self.registers: list[int] = cast(list[int], registers)
-        for i, value in enumerate(registers):
+        self.registers: list[int] = []
+        # copy values to not mutate the original sequence of values
+        for value in registers:
             if isinstance(value, bytes):
-                self.registers[i] = int.from_bytes(value, byteorder="big")
-        self.count: int = count or len(registers)
+                self.registers.append(int.from_bytes(value, byteorder="big"))
+            elif isinstance(value, int):
+                self.registers.append(value)
+            else:
+                raise ValueError("Invalid value in iterable")
+        self.count: int = count or len(self.registers)
         self.status: int = status
         self.fut: asyncio.Future
```

```shell
vfazio@vfazio4 /mnt/development/pymodbus $ cat test.py 
from pymodbus.client import ModbusTcpClient

mbc = ModbusTcpClient("localhost")

z: list[int] = [1, 2]

mbc.write_registers(10, z)

vfazio@vfazio4 /mnt/development/pymodbus $ mypy test.py 
test.py:7: error: Argument 2 to "write_registers" of "ModbusClientMixin" has incompatible type "list[int]"; expected "list[bytes | int]"  [arg-type]
test.py:7: note: "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
test.py:7: note: Consider using "Sequence" instead, which is covariant
Found 1 error in 1 file (checked 1 source file)
vfazio@vfazio4 /mnt/development/pymodbus $ git stash pop
On branch dev
Your branch is up to date with 'origin/dev'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   pymodbus/client/mixin.py
        modified:   pymodbus/pdu/pdu.py

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        build/lib/
        test.py

no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (75b0cf37e3435354a7f5c6911224ff32831bfbca)
vfazio@vfazio4 /mnt/development/pymodbus $ mypy test.py 
Success: no issues found in 1 source file

vfazio@vfazio4 /mnt/development/pymodbus $ mypy pymodbus
Success: no issues found in 47 source files
```

The mypy checks are not _complete_. I mentioned this in my previous comment

The test functions are not typed properly, so they are defaulting to the `Any` type which allows for any typing to pass.

When the proper typing is applied, they fail:

```diff
(venv) vfazio@vfazio4 /mnt/development/pymodbus $ git diff
diff --git a/examples/client_async_calls.py b/examples/client_async_calls.py
index d2c118e5..6dad4a95 100755
--- a/examples/client_async_calls.py
+++ b/examples/client_async_calls.py
@@ -34,6 +34,7 @@ import logging
 import sys
 
 from pymodbus import ModbusException
+from pymodbus.client import ModbusBaseClient
 from pymodbus.pdu import FileRecord
 
 
@@ -157,7 +158,7 @@ async def async_handle_holding_registers(client):
     assert not rr.isError()  # test that call was OK
     assert rr.registers == arguments["values"]
 
-async def async_write_registers_mypy(client):
+async def async_write_registers_mypy(client: ModbusBaseClient) -> None:
     """Read/write holding registers."""
     regs1: list[int] = [10] * 8
     await client.write_registers(1, regs1, slave=SLAVE)
```

```shell
(venv) vfazio@vfazio4 /mnt/development/pymodbus $ mypy examples/client_async_calls.py 
examples/client_async_calls.py:164: error: Argument 2 to "write_registers" of "ModbusClientMixin" has incompatible type "list[int]"; expected "list[bytes | int]"  [arg-type]
examples/client_async_calls.py:164: note: "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
examples/client_async_calls.py:164: note: Consider using "Sequence" instead, which is covariant
examples/client_async_calls.py:170: error: Argument 2 to "write_registers" of "ModbusClientMixin" has incompatible type "list[bytes]"; expected "list[bytes | int]"  [arg-type]
examples/client_async_calls.py:170: note: "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
examples/client_async_calls.py:170: note: Consider using "Sequence" instead, which is covariant
examples/client_async_calls.py:173: error: Non-overlapping equality check (left operand type: "list[int]", right operand type: "list[bytes]")  [comparison-overlap]
Found 3 errors in 1 file (checked 1 source file)
```

--- a/examples/client_async.py
+++ b/examples/client_async.py
@@ -48,7 +48,7 @@ _logger = logging.getLogger(__file__)
 _logger.setLevel("DEBUG")
 
 
-def setup_async_client(description=None, cmdline=None):
+def setup_async_client(description: str | None =None, cmdline: str | None = None) -> modbusClient.ModbusBaseClient:
     """Run client setup."""
     args = helper.get_commandline(
         server=False, description=description, cmdline=cmdline

--- a/examples/client_async_calls.py
+++ b/examples/client_async_calls.py
@@ -34,6 +34,7 @@ import logging
 import sys
 
 from pymodbus import ModbusException
+from pymodbus.client import ModbusBaseClient
 from pymodbus.pdu import FileRecord
 
 
@@ -157,19 +158,17 @@ async def async_handle_holding_registers(client):
     assert not rr.isError()  # test that call was OK
     assert rr.registers == arguments["values"]
 
-async def async_write_registers_mypy(client):
+async def async_write_registers_mypy(client: ModbusBaseClient) -> None:
     """Read/write holding registers."""
     regs1: list[int] = [10] * 8
     await client.write_registers(1, regs1, slave=SLAVE)
     rr = await client.read_holding_registers(1, count=len(regs1), slave=SLAVE)
     assert not rr.isError()  # test that call was OK
-    assert rr.registers == regs1
 
     regs2: list[bytes] = [b'\x01\x02', b'\x03\x04']
     await client.write_registers(1, regs2, slave=SLAVE)
     rr = await client.read_holding_registers(1, count=len(regs2), slave=SLAVE)
     assert not rr.isError()  # test that call was OK
-    assert rr.registers == regs2
 
 
 async def async_handle_input_registers(client):

--- a/examples/helper.py
+++ b/examples/helper.py
@@ -9,6 +9,7 @@ from __future__ import annotations
 import argparse
 import logging
 import os
+from typing import Any
 
 from pymodbus import pymodbus_apply_logging_config
 
@@ -16,7 +17,7 @@ from pymodbus import pymodbus_apply_logging_config
 _logger = logging.getLogger(__file__)
 
 
-def get_commandline(server=False, description=None, extras=None, cmdline=None):
+def get_commandline(server: bool = False, description: str | None = None, extras: Any = None, cmdline: str | None = None):
     """Read and validate command line arguments."""
     parser = argparse.ArgumentParser(description=description)
     parser.add_argument(

--- a/pymodbus/client/mixin.py
+++ b/pymodbus/client/mixin.py
@@ -2,6 +2,7 @@
 from __future__ import annotations
 
 import struct
+from collections.abc import Sequence
 from enum import Enum
 from typing import Generic, TypeVar
 
@@ -501,7 +502,7 @@ class ModbusClientMixin(Generic[T]):  # pylint: disable=too-many-public-methods
     def write_registers(
         self,
         address: int,
-        values: list[bytes | int],
+        values: Sequence[int | bytes],
         *,
         slave: int = 1,
         no_response_expected: bool = False

--- a/pymodbus/client/mixin.py
+++ b/pymodbus/client/mixin.py
@@ -516,7 +516,9 @@ class ModbusClientMixin(Generic[T]):  # pylint: disable=too-many-public-methods
         :raises ModbusException:
 
         .. tip::
-            values= entries defined as bytes are silently converted to int !
+            values= parameter:
+            entries defined as bytes are silently converted to int !
+            only list[int], list[bytes] or list[bytes | int] are expected (others may work unsupported)
 
         This function is used to write a block of contiguous registers
         (1 to approx. 120 registers) in a remote device.
@@ -628,7 +630,9 @@ class ModbusClientMixin(Generic[T]):  # pylint: disable=too-many-public-methods
         :raises ModbusException:
 
         .. tip::
-            values= entries defined as bytes are silently converted to int !
+            values= parameter:
+            entries defined as bytes are silently converted to int !
+            only list[int], list[bytes] or list[bytes | int] are expected (others may work unsupported)
 
         This function performs a combination of one read operation and one
         write operation in a single MODBUS transaction. The write

--- a/pymodbus/datastore/simulator.py
+++ b/pymodbus/datastore/simulator.py
@@ -106,10 +106,10 @@ class Setup:
     :meta private:
     """
 
-    def __init__(self, runtime):
+    def __init__(self, runtime: Any) -> None:
         """Initialize."""
         self.runtime = runtime
-        self.config = {}
+        self.config: Any = {}
         self.config_types: dict[str, dict[str, Any]] = {
             Label.type_bits: {
                 Label.type: CellType.BITS,

--- a/pymodbus/pdu/pdu.py
+++ b/pymodbus/pdu/pdu.py
@@ -4,6 +4,7 @@ from __future__ import annotations
 import asyncio
 import struct
 from abc import abstractmethod
+from collections.abc import Sequence
 from typing import cast
 
 from pymodbus.exceptions import NotImplementedException
@@ -24,7 +25,7 @@ class ModbusPDU:
             address: int = 0,
             count: int = 0,
             bits: list[bool] | None = None,
-            registers: list[int] | list[bytes] | list[bytes | int] | None = None,
+            registers: Sequence[bytes | int] | None = None,
             status: int = 1,
         ) -> None:
         """Initialize the base data for a modbus request."""

